Merge API Call HTTP Get options to a single struct

t0mk opened the pull request:
This fixes #207 #208 

@displague please take a look. Let's contine our converasion from #207 in here.

I moved the Options struct and logic to a new file. There were some changes in the functions, and I fixed it only in the fucntions in devices.go so far. I will do the rest after we agree on the format.

Some notes
* this will need a lot of testing, many not-so used listing funcs will be changed, and the correctness of those changes is sometimes not verifiable by the compiler. E.g. this leaves a lot of space for a bug that will only show at runtime:
```diff
-       opts = opts.Including("facility")
-       endpointPath := fmt.Sprintf("%s/%s%s", projectBasePath, projectID, deviceBasePath)
-       path := opts.WithQuery(endpointPath)
+func (s *DeviceServiceOp) List(projectID string, listOpt *ListOptions) (devices []Device, resp *Response, err error) {
+       listOpt = listOpt.Including("facility")
+       params := urlQuery(listOpt)
+       path := fmt.Sprintf("%s/%s%s?%s", projectBasePath, projectID, deviceBasePath, params)
```
* I didn't see the need to use the QueryAppender interface that youv'e suggested.
* The `Including` function creates a copy of the options struct. I think this is too implicit and unexpected. Also, if an API listing function won't call `Including` the first thing, the options (passed as a pointer) might be changed inside the API function and are no longer safe for reuse, like you suggested in #207 in
```
opts := &packngo.GetOptions{}
opts.Including("project", "facility", "plan")

leader, _, _ := client.Devices.Get(deviceID, opts)
followers, _, _ := client.Devices.List(projectID, opts)
```

displague commented:
> The Including function creates a copy of the options struct ... if an API listing function won't call Including the first thing, the options (passed as a pointer) might be changed inside the API function and are no longer safe for reuse

This was intentional because I wanted a function that guaranteed a query object with "?include=foo", so if the user sent `nil` options, and the function demands "?include=facility", the function can safely get that. If the user doesn't want their  pointer replaced, they should use a different variable name. I tried to express this in the documentation for the function, maybe we could improve that or include an example function.

This will serve as test coverage for the functions too. https://blog.golang.org/examples

```go
func ExampleGetOptions_Including_nil_caller() {
  var opt *GetOptions
  optCopy := opt.Including("facility")

  fmt.Println(opt)
  fmt.Println(optCopy.Includes)
  // Output:
  // nil
  // ["facility"]
}

func ExampleGetOptions_Including_zero_caller() {
  var opt := &GetOptions{}
  optCopy := opt.Including("devices", "users")

  fmt.Println(opt.Includes)
  fmt.Println(optCopy.Includes)
  // Output:
  // []
  // ["devices", "users"]
}
```

The only place that packngo uses `Including` itself, is in `List` and `Get` functions. As a user, I wouldn't want `Get` and `List` to change the options that I sent it.

I understand your expectation. Maybe the list function could create a copy with a separate function and do something like this:

```golang
func (DeviceService *ds) List(string, opt *GetOptions) (*[]Device, http.Response, error) {
  copyOpt := opt.CopyOrNew()
  copyOpt.Including("facility")
  ...
}
```

displague commented:
> The only place that packngo uses Including itself, is in List and Get functions. As a user, I wouldn't want Get and List to change the options that I sent it.

On the other hand...

It would be convenient for the Options I supply to be updated with the pagination data returned.

func (DeviceService *ds) List(string, opt *GetOptions) (*[]Device, http.Response, error) {
  copyOpt := opt.CopyOrNew()
  copyOpt.Including("facility")
  ...
  if opt != nil {
    opt.Pages = 15
    // any other metadata?
  }
}

displague commented on api_call_options.go:
```diff
@@ -94,26 +98,42 @@ func (g *GetOptions) GetPage() int { // guaranteed int
 	return g.Page
 }
 
+func (g *GetOptions) GetCopy() *GetOptions {
```
What do you think about `GetOptions() GetOptions`?  I can't help but think of https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#GetOptions and how the comparison differs because in those cases a DeepCopy does does not guarantee a GetOptions if the caller is nil.

The guarantee function we are making here is more like a `*string` to `string` getter:

```go
func (c *caller) GetName() String {
  if c == nil { return "" }
  if c.name == nil { return "" }
  return *c.name  
}
```

Except... We are both taking and returning a pointer. 

I was originally thinking `CopyOrNew` would be a more explicit name, but I can get behind `GetCopy`.
We should definitely include a comment describing this function as guaranteeing a  GetOptions response, which is either a copy of the caller or a new GetOptions if the caller is nil.

t0mk replied:
Thinking about it, from `GetCopy`, `GetOptions` and `CopyOrNew`, I would choose `CopyOrNew` as that's saying exactly what it does.

t0mk replied:
.. I wrote GetCopy because I needed it in func nextPage for forming the next-page URL with other opts (not straightforward because the meta.Next href will contain sth like `/projects?page=2`, no matter what opts you had in the original HTTP GET querystring). Considering we have to take the user-given opts from own variable, and we need to use the next page number, the next-page URL is not a great start for elegant URL building.

^ Maybe it could be suggested to the API devs: "Please include original querystring in the next-page href"

displague replied:
I created a ticket for that - common parameters (such as sort_by, sort_dir, includes, excludes, and search) should be reproduced in pagination links.

displague commented on api_call_options.go:
```diff
@@ -25,7 +26,7 @@ type GetOptions struct {
 	// references deep. (For example, "memberships.projects" can be used in
 	// ListUsers.)
-	Includes []string `url:"includes,omitempty"`
+	Includes []string `url:"include,omitempty,comma"`
```
Good catch. https://pkg.go.dev/github.com/google/go-querystring/query#Values

t0mk replied:
I only noticed after I tested it :)

displague commented on api_call_options.go:
```diff
@@ -94,26 +98,42 @@ func (g *GetOptions) GetPage() int { // guaranteed int
 }
 
 // nextPage is common and extracted from all List functions
 func nextPage(meta meta, opts *GetOptions) (path string) {
```
Would we be bonkers to implement `nextPage` as a channel or `go func` that would allow for all of the pages to be requested in parallel and reassembled in the expected order? We'd likely have to manage the queue, perhaps our HTTP libraries already offer features for this.

We can always revisit that later, it was just a thought.

t0mk replied:
This a good idea, but maybe I'd leave it for another PR.

t0mk commented on devices.go:
```diff
@@ -331,26 +331,21 @@ type DeviceServiceOp struct {
-	path := fmt.Sprintf("%s/%s%s?%s", projectBasePath, projectID, deviceBasePath, params)
+func (s *DeviceServiceOp) List(projectID string, opts *ListOptions) (devices []Device, resp *Response, err error) {
+	opts = opts.Including("facility")
+	apiCallUrl := getApiCallUrl(opts, projectBasePath, projectID, deviceBasePath)
```
@displague what do you think about this approach to build the API call URLs?

displague replied:
It seems like a convenient short-hand, but there is a downside to this. We won't have compile-time checks when the URL format string does not match the parameter listed.

My argument here is weak because we don't have a central body of URL paths and formats (constants representing each URL).

Here's an alternative to consider (and hopefully draw a better pattern from).

```
s.projectURL(opts, projectID)
s.projectURLWithIdentifier(opts, projectID, deviceID, action)
s.projectURLWithIdentifierAction(opts, projectID, deviceID, "action")
```

I'm not sold on this either, especially because we would have to copy this pattern to each service.

There are other approaches we could take that might minimize the amount of copying, possibly with an internal url helper package that defines the URL constants.

Do you have any thoughts on this @detiber?

displague replied:
@t0mk maybe we could open this last URL building change in a new PR. I'd like see the user-visible change introduced first and then we can optimize the internals.  I don't see a great deal of benefit in the change and I think it introduces risk (and I don't know if we can come up with a good alternative that we can agree on without delaying this PR).

Were we ready to remove the WIP tag?

t0mk replied:
@displague 

> s.projectURL(opts, projectID)
> s.projectURLWithIdentifier(opts, projectID, deviceID)
> s.projectURLWithIdentifierAction(opts, projectID, deviceID, "action")

I don't think this would make the code look very nice. I think there's not just 3 formats of the API endpoints in the Eq Metal API, so the API Call methods would not look very consistent. IMO just one general variadic function is the better looking alternativ.

But OK, I will not do the GetApiUrl in this PR.

I will get devices and volumes to the state of using the new general struct for options, and then ask you for review again. Once you ack that, I will reformat the rest of Packngo to using the new opt struct (wihout GetApiUrl).

t0mk commented on api_call_options.go:
```diff
@@ -0,0 +1,162 @@
+// getApiCallUrl will construct URL path with variables. It will build the
+// URL path from segments, and encode the options. If there are no options
+// in the API call, pass nil for opts.
+func getApiCallUrl(opts *GetOptions, pathSegments ...string) string {
```
@displague I added this whapper to further simplify the API call URL assembly in the packngo functions.

displague replied:
replied in https://github.com/packethost/packngo/pull/211#discussion_r526347547

t0mk commented:
@displague this is how I would like to merge this PR. In addition to merging the GetOpts to one struct, I replaced fmt.Sprintf with path.Join, and used GetOpts.WithQuery instead of urlQuery where possible. I did the former because I had to do it in some of the paginating List function, because of how the next page link is created and how the opts are treated there. And since I replaced fmt.Sprintf in one place, I thought I will go it everywhere. It was a lot of menial work, but it had to be done at some point. Sorry this PR is so big.

Since there was a lot of change in free-form string building, I would like to verify as much tests as possible, to check that nothing got broken.

Please take a look, and I will let you know once I test enough and I'm confident to give this a green light to merge.

displague commented on go.mod:
```diff
@@ -4,6 +4,7 @@ go 1.13
 
 require (
 	github.com/dnaeon/go-vcr v1.0.1
+	github.com/google/go-querystring v1.0.0
```
I'm more cognizant of this module import now that we've removed all other imports (if you exclude test-only imports).

displague replied:
We are only using this for `query.Values`.  Perhaps we can copy that function or find a standard-lib alternative?

t0mk replied:
`Values()` is the only exported function from google/go-querystring, and it's using all the other code from the lib, so copying code out is not really an option:
https://github.com/google/go-querystring/blob/master/query/encode.go#L113

I will take a look what would it take to do this with url.Values or something standard.

t0mk replied:
https://golang.org/pkg/net/url/#Values

t0mk replied:
@displague I removed the google/go-querystring dep. I will run some tests still, but you can check the code.

t0mk replied:
Tests succeed.

displague reviewed (approved):
I ran packet-cli with this branch via `go.mod` `replace`. Pagination, search, and include/exclude behavior worked as expected and there were no API compat issues to address in the CLI client.
